CS3219-AY2223S1/cs3219-project-ay2223s1-g47
# Rebuild user service and integrate with Docker

## Overview

Closes #45 

Main goals of the PR: 
- [x] Refactor user-service to typescript, and change organization to include exception handling, appropriate use of `.env` files, and abstraction from MongoDB
- [x] Fix bugs in JWT authentication
- [x] Add testing framework `jest` to user service
- [x] Set up population of `users` table in mongo DB (local and docker) when `ENV` is set to `DEV` 
- [x] Set up dockerization with specified `Dockerfile` and `docker-compose.yml`
- [x] Integrate frontend with user-service, testing on both local and docker

### What's next (in order)
- [x] Reorganize issues and sprint planning
- [ ] Add tests in `jest` for API calls with `supertest` 
- [ ] Clean up CI 
- [ ] Implement collaboration service
- [ ] Spruce up FE 
- [ ] Get matching service working @Gernene 
- [ ] Figure out deployment

### Testing

1. Check that `jest` testing works for you 

In `user-service/`, run `npm test`. 

2. Check that the frontend can integrate with the backend, with dummy data populated from `user-service/src/dev-data.json`

First, I recommend installing [MongoDB Compass (GUI)](https://zellwk.com/blog/local-mongodb/#:~:text=To%20connect%20to%20your%20local,databases%20in%20your%20local%20MongoDB.), a GUI for interfacing with the underlying tables. For an overview of important fundemental concepts to MongoDB, I recommend this [FAQ in the docs](https://www.mongodb.com/docs/manual/faq/fundamentals/#:~:text=Instead%20of%20tables%2C%20a%20MongoDB,in%20a%20relational%20database%20table.).

Use the `mongodb://localhost:27017` URI to connect. 
![image](https://user-images.githubusercontent.com/67091469/193445402-1aa9bb6b-9882-4722-8832-66aa374cbb15.png)


Then, run the containers with `docker-compose up`. 

You should see three containers - the frontend (at `localhost:3000`), the user service (at `localhost:8000`) and mongodb (at `mongodb://localhost:27017`). 

Now, signup with a new account. 
![image](https://user-images.githubusercontent.com/67091469/193445462-8da36832-dbbb-4382-9ed7-87d35788412c.png)

Then login with the wrong password or username. You should see an error snackbar.

![image](https://user-images.githubusercontent.com/67091469/193445482-ac6c2a39-0517-4987-ab03-1347a6acb9f0.png)

Then now login with the correct password. You will be redirected. When you try to refresh, observe that you stay logged in.

![image](https://user-images.githubusercontent.com/67091469/193445501-bf130173-e03a-4f4e-8483-42ed24b865a9.png)

To logout, since we don't have a button, simply delete the cookies in your browser.
